MDEV-24813 Signal full scan to storage engines.

mariadb-YuchenPei opened the pull request:
This is an initial patch for discussion.

Will work in the following minimal example

--source include/have_innodb.inc
CREATE TABLE t (a INT PRIMARY KEY) ENGINE=InnoDB;
insert into t values (42);

BEGIN;
SELECT * FROM t LOCK IN SHARE MODE; # will acquire S lock 
COMMIT;

DROP TABLE t;

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/not_signed)](https://cla-assistant.io/MariaDB/server?pullRequest=3487) <br/>Thank you for your submission! We really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla-assistant.io/MariaDB/server?pullRequest=3487) before we can accept your contribution.<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.io/check/MariaDB/server?pullRequest=3487) it.</sub>

dr-m commented on storage/innobase/include/row0mysql.h:
```diff
@@ -686,6 +686,8 @@ struct row_prebuilt_t {
 	/** The MySQL table object */
 	TABLE*		m_mysql_table;
 
+    bool full_table_lock=false;
```
The declaration of an initializer expression would introduce an implicit constructor `row_prebuilt_t::row_prebuilt_t()`. That in turn could lead into trouble with the GCC 6 `-flifetime-dse`, because the rest of the fields would not be zero-initialized.

All fields will be zero-initialized in `row_create_prebuilt()` by invoking `mem_heap_zalloc()`.

The field name is misleading: it should be `full_table_scan`. It is also missing a preceding Doxygen comment that explains when the flag would be set and what its impact is.

mariadb-YuchenPei replied:
Done.

mariadb-YuchenPei commented on sql/records.cc:
```diff
@@ -338,8 +338,8 @@ bool init_read_record(READ_RECORD *info,THD *thd, TABLE *table,
-      select && select->cond && 
-      (select->cond->used_tables() & table->map) &&
+      select && // select->cond &&
+      // (select->cond->used_tables() & table->map) &&
```
No longer commenting out these in the new patch

mariadb-YuchenPei commented:
> With this patch, we have "Using where with pushed condition" all over the place:

No longer the case with the new patch that does not comment out the cond_push if condition

mariadb-YuchenPei committed:
MDEV-24813 Signal full scan to storage engines.

When starting to do a full table/index scan without a WHERE or JOIN
condition, tell the storage engine so and the corresponding
ulong-truncated LIMIT.

Include an innodb implementation: added an innodb switch
table_lock_on_full_scan, so that when the switch is on, on receiving
the full scan signal from the sql layer, if the truncated LIMIT is
ULONG_MAX (likely no LIMIT), attempt to acquire a table lock.

Updated tests that have different results with the switch on.

(Comment and code edited by Sergei Petrunia)

Thirunarayanan committed:
MDEV-24813 Drop low-value table_lock test combinations

The three deadlock_*_race tests cannot reach their DEBUG_SYNC race
under a table lock (it degenerates to a timeout), and the three I_S
tests only restate a lock-mode change already covered
by innodb_full_scan.test.

- Addressed review comments

iMineLink commented on mysql-test/main/innodb_full_scan.test:
```diff
@@ -0,0 +1,211 @@
+SELECT * FROM t for update;
+COMMIT;
+--echo ## +0, was +1
+eval $print_lock_rec_lock_created;
```
I wonder if this can be changed so that it captures the delta from the previous "call"

iMineLink commented on mysql-test/main/innodb_full_scan.test:
```diff
@@ -0,0 +1,211 @@
+--echo #
+
+set @old_innodb_table_lock_on_full_scan=@@innodb_table_lock_on_full_scan;
+set innodb_table_lock_on_full_scan=1;
```
Can this be tested with combinations instead, so that the `.rdiff` shows the decrease in `lock_rec_lock_created` counts?

iMineLink commented on storage/innobase/handler/ha_innodb.cc:
```diff
@@ -15940,6 +15944,31 @@ ha_innobase::extra(
+ha_innobase::extra_opt(
+/*===============*/
+	enum ha_extra_function operation,
+			   /*!< in: HA_EXTRA_FLUSH or some other flag */
```
Maybe stale comment?

iMineLink commented on sql/sql_select.cc:
```diff
@@ -25338,6 +25338,7 @@ int test_if_use_dynamic_range_scan(JOIN_TAB *join_tab)
 int join_init_read_record(JOIN_TAB *tab)
 {
   bool need_unpacking= FALSE;
+  /* TODO: s/tab->join/join/g in this function */
```
What is this about?

mariadb-YuchenPei replied:
just a minor cleanup todo, nothing to do with innodb

iMineLink commented on storage/innobase/handler/ha_innodb.cc:
```diff
@@ -864,6 +864,10 @@ static MYSQL_THDVAR_BOOL(ft_enable_stopword, PLUGIN_VAR_OPCMDARG,
   /* default */ TRUE);
 
+static MYSQL_THDVAR_BOOL(table_lock_on_full_scan, PLUGIN_VAR_OPCMDARG,
+  "Whether to acquire table lock when SQL layer advises full scan",
```
Maybe add a reference to the fact that table lock is taken in place of row-locks for each row, so there is performance gain at the cost of less granular locking?

iMineLink reviewed (commented):
Please check my few minor comments, especially the ones around `innodb_full_scan.test`, thanks.

iMineLink commented:
Thanks for addressing my review comments, LGTM.

dr-m commented on mysql-test/suite/innodb/r/autoinc_debug,table_lock.rdiff:
```diff
@@ -0,0 +1,70 @@
+--- /home/ycp/source/mariadb-server/push-main/src/mysql-test/suite/innodb/r/autoinc_debug.result	2026-06-17 14:13:34.462383376 +1000
++++ /home/ycp/source/mariadb-server/push-main/src/mysql-test/suite/innodb/r/autoinc_debug,table_lock.reject	2026-06-17 16:11:38.709757127 +1000
```
Please follow the style of 88d9348dfc494105046451beef15a8e68d7f51be by omitting timestamps as well as absolute path names.

mariadb-YuchenPei replied:
sure that can be done, but why?

mariadb-YuchenPei replied:
done
